File: native/src/dxinterop.cpp
#include <jni.h>
#include <d3d9.h>
#include "openglfx.h"

// Emulate internal JavaFX's code for memory mapping
struct IManagedResource {
    void* virtualTable;

	IManagedResource* pPrev;
	IManagedResource* pNext;
};

struct D3DResource {
	IManagedResource managedResource;
	IDirect3DResource9* pResource;
	IDirect3DSwapChain9* pSwapChain;
	IDirect3DSurface9* pSurface;
	IDirect3DSurface9* pDepthSurface;
	IDirect3DTexture9* pTexture;

	D3DSURFACE_DESC desc;
};

wglChoosePixelFormatARBPtr       wglChoosePixelFormatARB;
wglCreateContextAttribsARBPtr    wglCreateContextAttribsARB;
wglDXOpenDeviceNVPtr             wglDXOpenDeviceNV;
wglDXCloseDeviceNVPtr            wglDXCloseDeviceNV;
wglDXRegisterObjectNVPtr         wglDXRegisterObjectNV;
wglDXSetResourceShareHandleNVPtr wglDXSetResourceShareHandleNV;
wglDXUnregisterObjectNVPtr       wglDXUnregisterObjectNV;
wglDXLockObjectsNVPtr            wglDXLockObjectsNV;
wglDXUnlockObjectsNVPtr          wglDXUnlockObjectsNV;

jlongArray createLongArray(JNIEnv* env, int size, jlong* elements) {
    jlongArray result = env->NewLongArray(size);
    env->SetLongArrayRegion(result, 0, size, elements);
    return result;
}

nvdxfun(jlong, hasNVDXInteropFunctions)(JNIEnv* env, jobject) {
    return wglDXOpenDeviceNV != 0;
}

nvdxfun(jlong, wglDXOpenDeviceNV)(JNIEnv* env, jobject, jlong dxDevice) {
    return (jlong)wglDXOpenDeviceNV((void*)dxDevice);
}

nvdxfun(jboolean, wglDXCloseDeviceNV)(JNIEnv* env, jobject, jlong hDevice) {
    return (jboolean)wglDXCloseDeviceNV((HANDLE)hDevice);
}

nvdxfun(jlong, wglDXRegisterObjectNV)(JNIEnv* env, jobject, jlong device, jlong dxResource, jint name, jint type, jint access) {
    return (jlong)wglDXRegisterObjectNV((HANDLE)device, (void*)dxResource, name, type, access);
}

nvdxfun(jboolean, wglDXSetResourceShareHandleNV)(JNIEnv* env, jobject, jlong dxResource, jlong shareHandle) {
    return (jboolean)wglDXSetResourceShareHandleNV((void*)dxResource, (HANDLE)shareHandle);
}

nvdxfun(jboolean, wglDXUnregisterObjectNV)(JNIEnv* env, jobject, jlong device, jlong object) {
    return (jboolean)wglDXUnregisterObjectNV((HANDLE)device, (HANDLE)object);
}

nvdxfun(jboolean, wglDXLockObjectsNV)(JNIEnv* env, jobject, jlong handle, jlong textureHandle) {
    return wglDXLockObjectsNV((HANDLE)handle, 1, (HANDLE*)&textureHandle);
}

nvdxfun(jboolean, wglDXUnlockObjectsNV)(JNIEnv* env, jobject, jlong handle, jlong textureHandle) {
    return wglDXUnlockObjectsNV((HANDLE)handle, 1, (HANDLE*)&textureHandle);
}

/*===========
      D3D
  ===========*/
nvdxfun(jlongArray, createD3DTexture)(JNIEnv* env, jobject, jlong _device, jint width, jint height) {
    IDirect3DDevice9Ex* device = (IDirect3DDevice9Ex*)_device;

    // It is important to set NULL
    IDirect3DTexture9* texture = NULL;
    HANDLE sharedHandle = NULL;
    device->CreateTexture(width, height, 0, D3DUSAGE_DYNAMIC, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, &texture, &sharedHandle);

    jlong array[] = { (jlong)texture, (jlong)sharedHandle };
    return createLongArray(env, 2, array);
}

nvdxfun(void, replaceD3DTextureInResource)(JNIEnv* env, jobject, jlong _resource, jlong newTexture) {
    D3DResource* resource = (D3DResource*)_resource;
    IDirect3DTexture9* texture = (IDirect3DTexture9*)newTexture;

    resource->pTexture->Release();
    resource->pResource->Release();
    resource->pSurface->Release();

    // From D3DResource in D3DResourceManager.cpp
    resource->pResource = texture;
    resource->pResource->AddRef();
    resource->pTexture = (IDirect3DTexture9*)resource->pResource;
    resource->pTexture->GetSurfaceLevel(0, &resource->pSurface);
    resource->pSurface->GetDesc(&resource->desc);
}

File: core/src/main/kotlin/com/huskerdev/openglfx/utils/windows/DXInterop.kt
package com.huskerdev.openglfx.utils.windows

import com.huskerdev.ojgl.GLContext
import com.huskerdev.openglfx.GLExecutor


const val WGL_ACCESS_WRITE_DISCARD_NV = 0x2

class DXInterop {
    companion object {

        @JvmStatic external fun wglDXOpenDeviceNV(dxDevice: Long): Long
        @JvmStatic external fun wglDXCloseDeviceNV(hDevice: Long): Boolean
        @JvmStatic external fun wglDXRegisterObjectNV(device: Long, dxResource: Long, name: Int, type: Int, access: Int): Long
        @JvmStatic external fun wglDXSetResourceShareHandleNV(dxResource: Long, shareHandle: Long): Boolean
        @JvmStatic external fun wglDXUnregisterObjectNV(device: Long, obj: Long): Boolean
        @JvmStatic external fun wglDXLockObjectsNV(interopHandle: Long, sharedTextureHandle: Long): Boolean
        @JvmStatic external fun wglDXUnlockObjectsNV(interopHandle: Long, sharedTextureHandle: Long): Boolean

        @JvmStatic external fun createD3DTexture(device: Long, width: Int, height: Int): LongArray
        @JvmStatic external fun replaceD3DTextureInResource(resource: Long, newTexture: Long)

        val interopHandle by lazy {
            var context = GLContext.current()
            return@lazy if(context.handle == 0L) {
                context = GLContext.create()
                context.makeCurrent()

                val result = createInteropHandle()
                GLContext.delete(context)
                result
            } else createInteropHandle()
        }

        private fun createInteropHandle(): Long {
            GLExecutor.initGLFunctions()
            return wglDXOpenDeviceNV(D3D9Device.fxInstance.handle)
        }

        fun isSupported() = interopHandle != 0L
    }
}